### PR TITLE
Shifted around calls relating to xrootd's BindRequest for clean cance…

### PR DESCRIPTION
**Pull request by jgates108**
…llation.

### CODE REVIEW

**Comment by fritzm** [comment]
Looks good to me, thanks!

File: core/modules/wcontrol/Foreman.cc [hunk 1]
```diff
@@ -71,23 +71,26 @@ Foreman::~Foreman() {
...
         int const resultProtocol = 2; // See proto/worker.proto Result protocol
         if(!msg.has_protocol() || msg.protocol() < resultProtocol) {
-            task->sendChannel->sendError("Unsupported wire protocol", 1);
+            LOGF_WARN("processMsg Unsupported wire protocol");
+            if (!task->getCancelled()) {
+                // We should not send anything back to xrootd if the task has been cancelled.
+                task->sendChannel->sendError("Unsupported wire protocol", 1);
+            }
         } else {
             auto qr = _newQueryRunner(task);
```
Comment (abh3) [inline core/modules/wcontrol/Foreman.cc]: I will always complain about the over-use of auto. The code looks "nice" but without my spending countless of hours I really have no idea what objects are being passed around or being returned. Which also adds a mysterious quality to the code. So, my review will likely be perfunctory.

File: core/modules/wdb/QueryRunner.cc [hunk 2]
```diff
@@ -269,7 +273,11 @@ void QueryRunner::_transmitHeader(std::string& msg) {
     // Make sure protoheader size can be encoded in a byte.
     assert(protoHeaderString.size() < 255);
     auto msgBuf = proto::ProtoHeaderWrap::wrap(protoHeaderString);
-    _task->sendChannel->sendStream(msgBuf.data(), msgBuf.size(), false);
+    if (!_cancelled) {
```
Comment (abh3) [inline core/modules/wdb/QueryRunner.cc]: I don't know how this flag is set but it likely should have been set using an atomic operation. This appears to be just a regular old inspection without any locks. So, how do you interlock testing cancellation with the act of cancellation. I don't see it.

File: core/modules/xrdsvc/SsiSession.cc [hunk 3]
```diff
@@ -74,19 +69,26 @@ void SsiSession::ProcessRequest(XrdSsiRequest* req, unsigned short timeout) {
...
-        replyChannel->sendError(os.str().c_str(), EINVAL);
-        ReleaseRequestBuffer();
+        errorFunc(os.str());
         return;
     }
 
     if(!(*_validator)(ru)) {
         LOGF_WARN("WARNING: unowned chunk query detected: %1%" % ru.path());
-        ReleaseRequestBuffer();
+        errorFunc("", false);
```
Comment (abh3) [inline core/modules/xrdsvc/SsiSession.cc]: No descriptive error message here?

File: core/modules/wcontrol/Foreman.cc (hunk 1 above)
Comment (jgates108) [inline core/modules/wcontrol/Foreman.cc]: It's a new QueryRunner, and _newQueryRunner is defined 10 lines below.

File: core/modules/wdb/QueryRunner.cc (hunk 2 above)
Comment (jgates108) [inline core/modules/wdb/QueryRunner.cc]: _cancelled is of type std::atomic<bool>.

File: core/modules/xrdsvc/SsiSession.cc (hunk 3 above)
Comment (jgates108) [inline core/modules/xrdsvc/SsiSession.cc]: I'll add one

**Pull request merged**